Support explicit image sizes in markdown

jurgenhaas opened the pull request:
#### Summary
Sometimes images have a size that's not suitable, like e.g. avatars when posting notifications from some tracking systems like github.com through bots. This PR allows to use image size defintion in markdown like this:

`![avatar](https://example.com/avatar/user.png =50x50)`

and converts that into width and height attributes in HTML output.

#### Ticket Link
http://forum.mattermost.org/t/how-to-define-size-of-images/2574

mattermod commented:
Thanks @jurgenhaas for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

jwilander commented:
Hi @jurgenhaas, thanks for the PR!

Looks like you have the following ESLint style errors, could you fix them and push the change?
```
/var/lib/jenkins/jobs/mattermost-platform-pr/workspace/src/github.com/mattermost/platform/webapp/utils/markdown.jsx
  121:13  error  'parts' is never reassigned. Use 'const' instead     prefer-const
  123:17  error  'lastPart' is never reassigned. Use 'const' instead  prefer-const
  124:13  error  Assignment to function parameter 'href'              no-param-reassign
  125:29  error  Expected '===' and instead saw '=='                  eqeqeq
  133:31  error  Expected '===' and instead saw '=='                  eqeqeq

✖ 5 problems (5 errors, 0 warnings)
```

You can check these yourself by running `make check-style`

jurgenhaas commented:
@jwilander sorry I missed that - corrected the errors immediately.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-221ebc8c.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-221ebc8c

jasonblais commented:
Thanks @jurgenhaas for the pull request! We've queued an item in [our UX channel to discuss it](https://pre-release.mattermost.com/core/pl/qtcajjgue78j5dko9w8p39s5ac).

You're welcome to join the discussion!

lfbrock commented:
@jurgenhaas - after discussing this, we're leaning towards keeping the aspect ratio if only width is set, but if both width and height are set resize the image to whatever was specified. 

The feeling was that if the user is choosing to specify both width and height, it makes sense to listen to what they set instead of overriding it to maintain the aspect ratio. Let me know your thoughts on this and if you're okay making that change? You can also message me on pre-release (`@lindsay`)

jurgenhaas commented:
The latest commit now supports all 4 versions:

1. No dimensions provided
2. Both dimensions provided such as `=20x30`
3. One dimension provided with final x such as `=20x`
4. One dimension provided without final x such as `=20`

The output for the versions 2, 3 and 4 is:

2. provide width and height attributes
3. provide height with value `auto`
4. only provide width attribute

Is that what you were looking for?

mattermod commented:
Spinmint test server destroyed

esethna commented:
Hi @jurgenhaas,

Just letting you know there may be a delay in reviewing and merging this PR as we are locking down feature and improvement contributions for the v3.6 release, per our [release process](https://docs.mattermost.com/process/release-process.html). This PR can be merged into `master` after we branch `release-3.6` on January 4th.

jurgenhaas commented:
Is there really no way to get this into 3.6? I mean it doesn't change the behavior for anyone, only for those who start using additional markup. So it can't break anything but would allow to get rid of those ugly big thumbnails that fill up the chat channel.

jasonblais commented:
Hey @jurgenhaas 

Really, really sorry for taking so long to review this PR -- some of our PM team has been away during the holidays and we're currently gearing up for the 3.6 release, so the review process is taking longer than usual.

Given our release candidate cut for 3.6 is tomorrow, we won't have time to get this into the upcoming release.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-08f6f8f221b34ed65.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-08f6f8f221b34ed65

jasonblais commented:
Sidhant is helping test this PR for us,

jasonblais commented:
Moved testing over to myself since it was taking a while

@jurgenhaas I did a quick smoke test just now, and it appears that if you specify both dimensions with `=20x30` for instance, the aspect ratio is still maintained (and it considers the first value for width only).

![image](https://cloud.githubusercontent.com/assets/13119842/21759108/b97fbe1c-d60f-11e6-8d7c-9718a3623193.png)

Can you help take a look?

jurgenhaas commented:
Thanks @jasonblais for testing this, I wasn't able to test that myself due to the lack of a testing environment.

What I can see from your example is this:

![2017-01-09 08 10 14 screen](https://cloud.githubusercontent.com/assets/1475675/21759257/2fb4bb54-d643-11e6-849c-b1d9d62d0490.jpg)

It's several things:

- Something is adding `style="height=auto"` to the html markup which overwrites the height attribute.
- If that was deleted, there is a height definition of 500px for the selector `#post-list .markdown-inline-img` in the CSS which I find confusing as this would rather apply for any image everywhere.

It looks like the auto height may have been added to fix the problem with the 500px height? Maybe someone has some knowledge about the history of the code here? If not, I get deeper into this trying to find out.

jasonblais commented:
@hmhealey Would you know?

hmhealey commented:
It's being added by the `markdownImageLoaded` function at the top of `webapp/utils/markdown.jsx` that's used to shrink the image after it's loaded. The reason that we start all images at 500px and then shrink them after they're loaded is because having them increase in size has historically caused the rest of the post list to shift downwards so that the user's screen appears to be scrolled up from the bottom.

You might be able to change it so that `markdownImageLoaded` is only called for images without their size explicitly specified. That should stop it from maintaining the aspect ratio.

jurgenhaas commented:
Thanks @hmhealey for the heads up on this. I've now set the height in the style attribute to the given height if available because if we omit the height style attribute completely, then the height of 500px would still apply.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0011e6b0f2a9a4b01.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0011e6b0f2a9a4b01

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-015b69c407a6100c5.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-015b69c407a6100c5

jasonblais commented:
@jurgenhaas looks like the aspect ratio is still maintained and is only based on the width parameter:

![image](https://cloud.githubusercontent.com/assets/13119842/21861584/7473093c-d803-11e6-9121-e5ae0d2c9c7a.png)

jurgenhaas commented:
I had committed a fix prior to your test but after the test environment being built. I guess that that fix is not deployed to that test environment yet, is it?

jasonblais commented:
Ah, sorry @jurgenhaas - I thought it was included with https://github.com/mattermost/platform/pull/4803/commits/f20d46f0fae0e36e6883a0bf96282640b502d520

Will re-build the test server, thanks!

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-085e5ca1ada4b7978.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-085e5ca1ada4b7978

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0c2a11665c6323376.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0c2a11665c6323376

jurgenhaas commented:
Great, now it's working as expected.

jasonblais commented:
Hey @jurgenhaas,

what @lfbrock meant above regarding the `=20` and `=20x` that is followed after an image URL is that she was hoping both of them would work the same way, i.e. they would set the width parameter while keeping the aspect ratio.

Sorry if that wasn't clear...would you be able to help with an update?

jurgenhaas commented:
@jasonblais git you, fixed that too.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-09db0a0ce65550684.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-09db0a0ce65550684

jurgenhaas commented:
I've tested it and think it is fully working now.

jasonblais commented:
Thanks @jurgenhaas, two questions:

1) It looks like when you specify `=20` or `=20x` only the width parameter is set correctly, but the aspect ratio isn't really maintained. 

Example shown below, can you help take a look?

![image](https://cloud.githubusercontent.com/assets/13119842/21959186/7c5d5ed6-da8d-11e6-98d9-0f1b555c3ae6.png)



2) It appears that if I set the height parameter to be greater than 500px, the height shrinks back down to 500px.

This is not really an issue as we'd want to have some max height set --> I'm just wondering if 500px is the maximum dimension for height so we can document it properly? From the earlier conversation it sounds like it would be.

jurgenhaas commented:
Yes, the max height of 500px had been defined even before this issue and is therefore somehow unrelated. But documenting that is a good idea.

In fact this max height explains both of the effects you're describing.

jasonblais commented:
Oh, got it, of course the max height is the cause of 1) :) 

For the case in 1), would the user expect that the aspect ratio is maintained though?  

Example 1:
  - image dimension: 20x100
  - user posts image with size `=200`
  - observed: image scaled up to 200x1000 but shrinked vertically to 200x500 since the max height is 500px --> should the image shrink both horizontally AND vertically to 100x500 in this case?

Example 2:  
  - user posts image with size `=200x1000` (so specified both width and height)
  - observed: image scaled up to 200x1000 but shrinked vertically to 200x500 since the max height is 500px --> should the image shrink both horizontally AND vertically to 100x500 in this case, to keep the specified aspect ratio?

What do you think?

jurgenhaas commented:
@jasonblais That's a good question, but that has been the case even before introducing this new feature to define the dimensions and people may have wondered before as well. Do if that's something to be addressed, wouldn't that be a separate issue?

jasonblais commented:
> but that has been the case even before introducing this new feature

I think previously we kept the aspect ratio --> if the image was too high vertically, we shrank the image both vertically and horizontally until the height was at the maximum.

I think it would make sense to do the same when the image size is specified but exceeds the vertical maximum dimension, so that the behaviour is consistent. If that's a lot of work, let me know.

jurgenhaas commented:
Right, I see where you're coming from. Without specified dimensions, the max-height of 500 always applied and to preserve the aspect ratio, the width got reduced if necessary.

Making sure we still keep aspect ration when just a width is defined, is possible by then unsetting the max height. However, that could cause a different problem: as @hmhealey explained above, the handling of images was built to set them to a fixed height of 500px originally and later remove that fix height and keep the max height. The result of that is that the canvas of the stream gets pretty hight and scrolls to the end and when the image is completely loaded from the remote source which could happen with a delay, it will then be that size or smaller. That makes sure that the user is certainly seeing the end of the stream.

If we now still start with the fix heigt of 500px and then leave it to the real dimensions of the image, we would run into exactly that problem, if the real image is going to be taller than that.

The problem is, that while putting the markup together, we don't know the real dimensions of the image and therefore can't react on it.

But what ever you want to specify here, implementation should be simple, as long as it's possible at all.

jasonblais commented:
Ah cool, let me chat with hmhealey on this today

jasonblais commented:
Hey @jurgenhaas, 

Would it be possible to use CSS to set the image's max height as opposed to its actual height? 

That way we might be able to cover the following case:

1. you have an image with dimensions 10x100 , then
2. specify the width to be 100 (with `100x` in the image URL), which would make the image size to be 100x1000
3. but before posting, have it shrink from 100x1000 to 50x500 to keep the aspect ratio while decreasing the height to a max of 500px

jurgenhaas commented:
Unfortunately, CSS doesn't have the capability to do that. You would need JavaScript to do that but that loads later than what you would require to keep that workaround for some clients that aren't able top handle the increase of the height after the content has been loaded initially. If we still want to keep it working for those clients too, and that's the only purpose of the whole max-height exercise, then we have to do it with something that's known right at the beginning, i.e. when loading the content. That excludes all of JavaScript as this is only executed after Document.ready().

jasonblais commented:
>  then we have to do it with something that's known right at the beginning, i.e. when loading the content

Would that be feasible? Not sure what's the technical work for it.

jurgenhaas commented:
Well, you don't know the image size before you've loaded it. And HTML is designed such that sich resources ate loaded afterwards. Thinking about it feels like trying to resolve something just because some client software doesn't operate like modern tools should. This energy is probably better used in fixing such clients.  Do we know more about the clients that are causing that problem?

hmhealey commented:
@jurgenhaas What if you were to do use the markdownImageLoaded to change the image's width after the browser has loaded it? Something like
```javascript
function markdownImageLoaded(image) {
    if (image.hasAttribute('height') && image.attributes.height.value !== 'auto') {
        const maxHeight = parseInt(global.getComputedStyle(image).maxHeight);

        if (image.attributes.height.value > maxHeight && image.hasAttribute('width')) {
            image.style.height = maxHeight + 'px';
            image.style.width = (maxHeight / image.attributes.height.value * image.attributes.width.value) + 'px';
        } else {
            image.style.height = image.attributes.height.value + 'px';
        }
    } else {
        image.style.height = 'auto';
    }
}
```
It'd be nicer to do with pure CSS, but since it's not possible that way, I think this'll work.

jurgenhaas commented:
Possible yes. But that will lead into the same situation that the canvas is getting larger/longer than before with the mentioned problems for some clients. If that were acceptable, then we could just remove the max height and get exactly the same solution in pure css.

jasonblais commented:
Hey @jurgenhaas, I think @hmhealey's code won't make the image's height ever exceed the max height, so it'll never become taller after it loads.

So the canvas shouldn't get any larger/longer than 500 px

jasonblais commented:
Hey @jurgenhaas, wondering if you had any thoughts?  We're definitely open to other suggestions as well if you've got any.

Really close that we've completed the improvement, this will be a nice one :)

jurgenhaas commented:
@jasonblais sorry I was out of town and only returned to my desk to have a look at this. I have included a modified version of @lfbrock code:

- I removed the check for the width attribute, because if and image has the height attribute it will definitely also have the width
- In the formula for the new height I have changed the order of operation: first multiply and then divide, because the other way round we could get into trouble if height is only slightly bigger than max height and the division might get rounded to zero on some platforms

When a new test built is going to be deployed we can have a look for this use case. But there is certainly another one which is not yet covered: if only the width is defined and not the height (i.e. if height is set to auto), then we don't know if we're going to be higher than max height and we have nothing to do any calculation with. I actually don't know how to get the real height of an image if the displayed height has been limited by max-height already.

jasonblais commented:
Thanks @jurgenhaas! And not a problem at all, was just checking to see if you had seen it.

Looks like there's a build error -- can you help take a look and I'll set up a new test server for this PR

jurgenhaas commented:
@jasonblais sorry, 2 of the formulas had lint issues. Just fixed and re-committed those.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-01be375587664d1a9.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-01be375587664d1a9

jurgenhaas commented:
Right, the given use case is working ok. Now only this one remains the problem: `![image](http://www.maniacworld.com/now-this-is-a-tall-building.jpg =100)` - the height should become something like 670px but gets maxed at 500px.

jasonblais commented:
Thanks @jurgenhaas, would this part not cover the case?

```
        if (image.attributes.height.value > maxHeight && image.hasAttribute('width')) {
            image.style.height = maxHeight + 'px';
            image.style.width = (maxHeight / image.attributes.height.value * image.attributes.width.value) + 'px';
        } else {
            image.style.height = image.attributes.height.value + 'px';
        }
```

Or is it that there isn't a way to retrieve the value of `image.attributes.height.value`

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0c163eefe4f1cef02.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0c163eefe4f1cef02

jurgenhaas commented:
@jasonblais when `![image](http://www.maniacworld.com/now-this-is-a-tall-building.jpg =100)` is given, then html will contain the attribute width equals 100 and height equals auto. There is no numeric value for height that you could use to do any math.

jasonblais commented:
But if `![image](http://www.maniacworld.com/now-this-is-a-tall-building.jpg)` is posted, the aspect ratio is maintained while the image height shrinks to 500px.

Why is that? Could we use a similar calculation when the image width is specified? Or is the calculation done differently in that case.

jurgenhaas commented:
In your example the width is not specified, that's why only max height is applied and the width gets adjusted. Bur when the width is specified and a max height, then you can't also comply with aspect ratio.

jasonblais commented:
Thanks @jurgenhaas, I see what you mean.

Okay, given that's an edge case, we can leave it as-is for now. It'll be much more difficult to fix, and it won't happen very often (in worst case, the user can edit the width parameter accordingly).

I re-tested the PR otherwise and it works great! Moving to dev review next.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0206d5c821a3b83a2.spinmint.com:8065/pr4803

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0206d5c821a3b83a2

jasonblais commented:
Thanks @jurgenhaas for your contribution! :tada:

jurgenhaas commented:
Thanks @jasonblais for your patient support.
